chia/wallet/wallet_singleton_store.py

```diff
@@ -0,0 +1,281 @@
+        return None
+
+    async def get_latest_singleton(
+        self, singleton_id: bytes32, only_confirmed: bool = False
+    ) -> Optional[SingletonCoinRecord]:
+        """
+        Check for a record with coin ID and return the most recent if present
+        """
+        async with self.db_wrapper.reader_no_transaction() as conn:
+            if only_confirmed:
+                cursor = await conn.execute(
+                    "SELECT * from singletons WHERE singleton_id=? and confirmed=1 ORDER BY generation DESC LIMIT 1",
+                    (singleton_id,),
+                )
+            else:
+                cursor = await conn.execute(
+                    "SELECT * from singletons WHERE singleton_id=? ORDER BY generation DESC LIMIT 1",
+                    (singleton_id,),
+                )
+            row = await cursor.fetchone()
+            await cursor.close()
+        if row is not None:
+            custom_data = await self.get_custom_data_by_coin_id(row[0])
+            return _row_to_singleton_record(row, custom_data)
+        return None
+
+    async def get_unconfirmed_singletons(self, singleton_id: bytes32) -> List[SingletonCoinRecord]:
+        """
+        Returns all singletons with a specific id that are unconfirmed
+        """
+        async with self.db_wrapper.reader_no_transaction() as conn:
+            rows = await conn.execute_fetchall(
+                "SELECT * from singletons WHERE singleton_id=? AND confirmed=0", (singleton_id,)
+            )
+        records = []
+        for row in rows:
+            custom_data = await self.get_custom_data_by_coin_id(row[0])
+            records.append(_row_to_singleton_record(row, custom_data))
+        return records
+
+    async def set_confirmed(self, coin_id: bytes32, confirmed_height: uint32, timestamp: uint64) -> None:
+        """
+        Updates a record to set confirmed=True, and add the confirmed height
+        """
+        current: Optional[SingletonCoinRecord] = await self.get_record_by_coin_id(coin_id)
+        if current is None or (current.confirmed_at_height == confirmed_height and current.confirmed):
+            return
+
+        await self.add_singleton_record(
+            dataclasses.replace(current, confirmed=True, confirmed_at_height=confirmed_height, timestamp=timestamp)
+        )
+
+        # update the parent record spent_height
+        parent_id = current.coin.parent_coin_info
+        parent: Optional[SingletonCoinRecord] = await self.get_record_by_coin_id(parent_id)
+        if parent is None:
+            return
+        await self.add_singleton_record(dataclasses.replace(parent, spent_height=confirmed_height))
+
+    async def set_spent(self, coin_id: bytes32, spent_height: uint32, timestamp: uint64) -> None:
+        """
+        Updates a confirmed record with a spent_height
+        """
+        current: Optional[SingletonCoinRecord] = await self.get_record_by_coin_id(coin_id)
```

Comment:
it's *very* slow to get the record, update it and then store it back. Just deserializing and serializing the class is probably a material cost.

I would think we would want a single SQL statement that updates these fields, specifically and then is done.

The same comment applies to `set_confirmedI()`